detox/src/devices/drivers/android/genycloud/GenyCloudDriver.js

```diff
@@ -0,0 +1,66 @@
+const _ = require('lodash');
+const AndroidDriver = require('../AndroidDriver');
+const GenyCloudExec = require('./exec/GenyCloudExec');
+const RecipesService = require('./services/GenyRecipesService');
+const InstanceLookupService = require('./services/GenyInstanceLookupService');
+const InstanceLifecycleService = require('./services/GenyInstanceLifecycleService');
+const InstanceNaming = require('./services/GenyInstanceNaming');
+const AllocationHelper = require('./helpers/GenyAllocationHelper');
+const DeviceQueryHelper = require('./helpers/GenyDeviceQueryHelper');
+const DetoxRuntimeError = require('../../../../errors/DetoxRuntimeError');
+const logger = require('../../../../utils/logger').child({ __filename });
+
+class GenyCloudDriver extends AndroidDriver {
+  constructor(config) {
+    super(config);
+    this._name = 'Unspecified Genymotion Cloud Emulator';
+
+    const exec = new GenyCloudExec();
+    const instanceNaming = new InstanceNaming(); // TODO should consider a permissive impl for debug/dev mode. Maybe even a custom arg in package.json (Detox > ... > genycloud > sharedAccount: false)
+    this.recipeService = new RecipesService(exec, logger);
+    this.instanceLookupService = new InstanceLookupService(exec, instanceNaming, this.deviceRegistry);
+    this.instanceLifecycleService = new InstanceLifecycleService(exec, instanceNaming);
+    this.deviceQueryHelper = new DeviceQueryHelper(this.recipeService);
+    this.allocationHelper = new AllocationHelper(this.instanceLookupService, this.instanceLifecycleService);
+  }
+
+  get name() {
+    return this._name;
+  }
+
+  async acquireFreeDevice(deviceQuery) {
+    const recipe = await this.deviceQueryHelper.getRecipeFromQuery(deviceQuery);
+    this.assertRecipe(deviceQuery, recipe);
+
+    const cookie = { coldBoot: false };
+    const adbName = await this.allocateDevice({
+      recipe,
+      cookie,
+      toString: () => recipe.toString(),
+    });
+
+    await this.emitter.emit('bootDevice', { coldBoot: cookie.coldBoot, deviceId: adbName, type: recipe.name});
+    await this.adb.apiLevel(adbName);
+    await this.adb.disableAndroidAnimations(adbName);
+
+    this._name = `GenyCloud-${adbName}`;
+    return adbName;
+  }
+
+  async doAllocateDevice({ recipe, cookie }) {
+    const { instance, coldBooted } = await this.allocationHelper.allocateInstance(recipe.uuid);
```

Comment:
Continuing my comment about the abstraction leak.
In other words, what you should have been doing, if you follow the current pattern of device allocation is:

```js
class AllocationHelper {
  allocateInstance(uuid) {
    let instance = await instanceLookupService.findFreeInstance(recipeUUID);
    if (instance) {
      return instance.someId;
    }
    
    // otherwise
    return instanceNaming.generateName();
  }
}
```

Apparently, you have hit some issue with those layers and their relations, so you have put everything inside `allocateInstance`. I don't like that because of two possible reasons:

1) that allocation workflow I had written previously, it just doesn't pass reality check – then we should refactor AttachedDriver and EmulatorDriver, and get rid of my naive code there;
2) you are misusing the current allocation workflow, which is pretty solid and allows implementing this flow in a more idiomatic way.

I don't really care whether it is 1 or 2 – as soon as we understand what's odd here, we'll remove that breach and the code will just get better by a mile.